src/foam/nanos/dig/SugarWebAgent.java

```diff
@@ -101,7 +101,7 @@ public void execute(X x) {
               Class paramObj_ = null;
               Object obj_ = null;
 
-              if ( ! SafetyUtil.isEmpty(p.getParameter(pArray[j].getName())) ) {
+              if ( ! SafetyUtil.isEmpty(p.getParameter(pArray[j].getName())) || typeName.equals("foam.core.X") ) {
```

Comment:
At first, it wasn't clear to me why this condition was being added here and I think it could warrant its own conditional block along with a comment explaining why. How does something like this look?
```
              if ( typeName.equals("foam.core.X") ) {
                // Contexts should not be passed in as arguments so force the use of
                // the current context when a foam.core.X argument is encountered.
                arglist[j] = x;
              } else if ( ! SafetyUtil.isEmpty(p.getParameter(pArray[j].getName())) ) {
                arglist[j] = setParameterType(x, resp, out, p, typeName, pArray[j]);
              } else {
```
If we do this, we can remove the `foam.core.X` case in the `setParameterType` method.

Somewhat unrelated: I was also being a bit thrown off by the name `setParameterType` because it's not really setting a type, it's setting a value based on the type. We might want to consider renaming that.

Reply:
Thank you, let me update when I have a chance asap.